File: app-android/src/main/java/com/ustadmobile/port/android/view/SelectMultiplePeopleFragment.java
package com.ustadmobile.port.android.view;

import android.app.Dialog;
import android.arch.lifecycle.LiveData;
import android.arch.paging.DataSource;
import android.arch.paging.LivePagedListBuilder;
import android.arch.paging.PagedList;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.support.annotation.DrawableRes;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.content.res.AppCompatResources;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.SelectMultiplePeoplePresenter;
import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.view.DismissableDialog;
import com.ustadmobile.core.view.SelectMultiplePeopleView;
import com.ustadmobile.lib.db.entities.Person;
import com.ustadmobile.port.android.util.UMAndroidUtil;

import java.util.HashMap;
import java.util.Objects;

import io.reactivex.annotations.NonNull;

/**
 * SelectPeopleDialogFragment Android fragment extends UstadBaseFragmentRe
 */
public class SelectMultiplePeopleFragment extends UstadDialogFragment implements
        SelectMultiplePeopleView,  DismissableDialog {


    AlertDialog dialog;
    View rootView;

    //RecyclerView
    private RecyclerView mRecyclerView;
    private RecyclerView.LayoutManager mRecyclerLayoutManager;
    private RecyclerView.Adapter mAdapter;

    private SelectMultiplePeoplePresenter mPresenter;
    //Context (Activity calling this)
    private Context mAttachedContext;

    boolean forActor;

    Toolbar toolbar;

    HashMap<String, Long> selectedPeople;

    //Main Activity should implement this ?
    public interface PersonSelectDialogListener{
        void onSelectPeopleListener(HashMap<String, Long> selected, boolean actor);
    }

    /**
     * Generates a new Fragment for a page fragment
     *
     * @return A new instance of fragment SelectClazzesDialogFragment.
     */
    public static SelectMultiplePeopleFragment newInstance() {
        SelectMultiplePeopleFragment fragment = new SelectMultiplePeopleFragment();
        Bundle args = new Bundle();
        fragment.setArguments(args);
        return fragment;
    }


    public Drawable getTintedDrawable(Drawable drawable, int color) {
        drawable = DrawableCompat.wrap(drawable);
        int tintColor = ContextCompat.getColor(getContext(), color);
        DrawableCompat.setTint(drawable, tintColor);
        return drawable;
    }

    @android.support.annotation.NonNull
    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        LayoutInflater inflater =
                (LayoutInflater) Objects.requireNonNull(getContext()).getSystemService(
                        Context.LAYOUT_INFLATER_SERVICE);
        assert inflater != null;

        rootView = inflater.inflate(R.layout.fragment_select_people_dialog, null);

        //Set up Recycler View
        initView();

        //Toolbar:
        toolbar = rootView.findViewById(R.id.fragment_select_people_dialog_toolbar);
        toolbar.setTitle(R.string.selected_les);

        Drawable upIcon = AppCompatResources.getDrawable(getContext(),
                R.drawable.ic_arrow_back_white_24dp);

        upIcon = getTintedDrawable(upIcon, R.color.icons);

        toolbar.setNavigationIcon(upIcon);
        toolbar.setNavigationOnClickListener(v -> dialog.dismiss());


        toolbar.inflateMenu(R.menu.menu_done);
        //Click the tick button on the toolbar:
        toolbar.setOnMenuItemClickListener(item -> {
            int i = item.getItemId();
            if (i == R.id.menu_done) {
                mPresenter.handleCommonPressed(-1);
            }
            return false;
        });

        mPresenter = new SelectMultiplePeoplePresenter(getContext(),
                UMAndroidUtil.bundleToHashtable(getArguments()), this);
        mPresenter.onCreate(UMAndroidUtil.bundleToHashtable(savedInstanceState));

        //Dialog stuff:
        //Set any view components and its listener (post presenter work)
        dialog = new AlertDialog.Builder(getContext(),
                R.style.FullScreenDialogStyle)
                .setView(rootView)
                .setTitle("")
                .create();
        return dialog;

    }


    public static Drawable getTintedDrawable(@NonNull final Context context,
                                             @DrawableRes int drawableRes, @ColorRes int colorRes) {
        Drawable d = ContextCompat.getDrawable(context, drawableRes);
        d = DrawableCompat.wrap(d);
        DrawableCompat.setTint(d.mutate(), ContextCompat.getColor(context, colorRes));
        return d;
    }

    private void initView(){
        //Set recycler view
        mRecyclerView = rootView.findViewById(R.id.fragment_select_people_dialog_recyclerview);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
    }

    @Override
    public void onAttach(Context context){
        super.onAttach(context);
        this.mAttachedContext = context;
        this.selectedPeople = new HashMap<>();
    }

    @Override
    public void onDetach(){
        super.onDetach();
        this.mAttachedContext = null;
        this.selectedPeople = null;
    }

    @Override
    public void finish(){
        selectedPeople = mPresenter.getPeople();
        if(mAttachedContext instanceof PersonSelectDialogListener){
             ((PersonSelectDialogListener) mAttachedContext).onSelectPeopleListener(selectedPeople,
                     forActor);
        }
        dialog.dismiss();
    }

    /**
     * The DIFF CALLBACK
     */
    public static final DiffUtil.ItemCallback<Person> DIFF_CALLBACK =
        new DiffUtil.ItemCallback<Person>() {
            @Override
            public boolean areItemsTheSame(Person oldItem,
                                           Person newItem) {
                return oldItem.getPersonUid() == newItem.getPersonUid();
            }

            @Override
            public boolean areContentsTheSame(Person oldItem,
                                              Person newItem) {
                return oldItem.equals(newItem);
            }
        };

    @Override
    public void setListProvider(UmProvider<Person> personListProvider) {
        PersonListReturnSelectedRecyclerAdapter mAdapter =
                new PersonListReturnSelectedRecyclerAdapter(DIFF_CALLBACK, getContext(),
                        this, mPresenter);
        //A warning is expected
        DataSource.Factory<Integer, Person> factory =
                (DataSource.Factory<Integer, Person>)
                        personListProvider.getProvider();
        LiveData<PagedList<Person>> data =
                new LivePagedListBuilder<>(factory, 20).build();

        data.observe(this, mAdapter::submitList);

        mRecyclerView.setAdapter(mAdapter);
    }

    // This event is triggered soon after onCreateView().
    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        // Setup any handles to view objects here

    }

    @Override
    public void dismiss() {

    }
}

File: app-android/src/main/kotlin/com/ustadmobile/port/android/view/SelectMultiplePeopleFragment.kt
package com.ustadmobile.port.android.view

import android.app.Dialog
import android.content.Context
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import androidx.annotation.ColorRes
import androidx.annotation.DrawableRes
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.content.res.AppCompatResources
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.DrawableCompat
import androidx.lifecycle.Observer
import androidx.paging.DataSource
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.toughra.ustadmobile.R
import com.ustadmobile.core.controller.SelectMultiplePeoplePresenter
import com.ustadmobile.core.impl.UMAndroidUtil
import com.ustadmobile.core.view.DismissableDialog
import com.ustadmobile.core.view.SelectMultiplePeopleView
import com.ustadmobile.lib.db.entities.Person
import java.util.*

/**
 * SelectPeopleDialogFragment Android fragment extends UstadBaseFragmentRe
 */
class SelectMultiplePeopleFragment : UstadDialogFragment(),
        SelectMultiplePeopleView, DismissableDialog {
    override val viewContext: Any
        get() = context!!


    internal lateinit var dialog: AlertDialog
    internal lateinit var rootView: View

    //RecyclerView
    private var mRecyclerView: RecyclerView? = null
    private val mRecyclerLayoutManager: RecyclerView.LayoutManager? = null
    private val mAdapter: RecyclerView.Adapter<*>? = null

    private var mPresenter: SelectMultiplePeoplePresenter? = null
    //Context (Activity calling this)
    private var mAttachedContext: Context? = null

    internal var forActor: Boolean = false

    internal lateinit var toolbar: Toolbar

    internal var selectedPeople: HashMap<String, Long>? = null

    //Main Activity should implement this ?
    interface PersonSelectDialogListener {
        fun onSelectPeopleListener(selected: HashMap<String, Long>?, actor: Boolean)
    }


    fun getTintedDrawable(drawable: Drawable?, color: Int): Drawable {
        var drawable = drawable
        drawable = DrawableCompat.wrap(drawable!!)
        val tintColor = ContextCompat.getColor(context!!, color)
        DrawableCompat.setTint(drawable!!, tintColor)
        return drawable
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {

        val inflater = context!!.getSystemService(
                Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater

        rootView = inflater.inflate(R.layout.fragment_select_people_dialog, null)

        //Set up Recycler View
        initView()

        //Toolbar:
        toolbar = rootView.findViewById(R.id.fragment_select_people_dialog_toolbar)
        toolbar.setTitle(R.string.selected_les)

        var upIcon = AppCompatResources.getDrawable(context!!,
                R.drawable.ic_arrow_back_white_24dp)

        upIcon = getTintedDrawable(upIcon, R.color.icons)

        toolbar.navigationIcon = upIcon
        toolbar.setNavigationOnClickListener { v -> dialog.dismiss() }


        toolbar.inflateMenu(R.menu.menu_done)
        //Click the tick button on the toolbar:
        toolbar.setOnMenuItemClickListener { item ->
            val i = item.itemId
            if (i == R.id.menu_done) {
                mPresenter!!.handleCommonPressed(-1)
            }
            false
        }

        mPresenter = SelectMultiplePeoplePresenter(context!!,
                UMAndroidUtil.bundleToMap(arguments), this)
        mPresenter!!.onCreate(UMAndroidUtil.bundleToMap(savedInstanceState))

        //Dialog stuff:
        //Set any view components and its listener (post presenter work)
        dialog = AlertDialog.Builder(context!!,
                R.style.FullScreenDialogStyle)
                .setView(rootView)
                .setTitle("")
                .create()
        return dialog

    }

    private fun initView() {
        //Set recycler view
        mRecyclerView = rootView.findViewById(R.id.fragment_select_people_dialog_recyclerview)
        mRecyclerView!!.layoutManager = LinearLayoutManager(context)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        this.mAttachedContext = context
        this.selectedPeople = HashMap()
    }

    override fun onDetach() {
        super.onDetach()
        this.mAttachedContext = null
        this.selectedPeople = null
    }

    override fun finish() {
        selectedPeople = mPresenter!!.people
        if (mAttachedContext is PersonSelectDialogListener) {
            (mAttachedContext as PersonSelectDialogListener).onSelectPeopleListener(selectedPeople,
                    forActor)
        }
        dialog.dismiss()
    }

    override fun setListProvider(factory: DataSource.Factory<Int, Person>) {
        val mAdapter = PersonListReturnSelectedSimpleRecyclerAdapter(DIFF_CALLBACK, context!!,
                this, mPresenter!!)
        //A warning is expected
        val data = LivePagedListBuilder(factory, 20).build()

        data.observe(this, Observer<PagedList<Person>> { mAdapter.submitList(it) })

        mRecyclerView!!.adapter = mAdapter
    }

    // This event is triggered soon after onCreateView().
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        // Setup any handles to view objects here

    }

    override fun dismiss() {

    }

    companion object {

        /**
         * Generates a new Fragment for a page fragment
         *
         * @return A new instance of fragment SelectClazzesDialogFragment.
         */
        fun newInstance(): SelectMultiplePeopleFragment {
            val fragment = SelectMultiplePeopleFragment()
            val args = Bundle()
            fragment.arguments = args
            return fragment
        }


        fun getTintedDrawable(context: Context,
                              @DrawableRes drawableRes: Int, @ColorRes colorRes: Int): Drawable {
            var d = ContextCompat.getDrawable(context, drawableRes)
            d = DrawableCompat.wrap(d!!)
            DrawableCompat.setTint(d!!.mutate(), ContextCompat.getColor(context, colorRes))
            return d
        }

        /**
         * The DIFF CALLBACK
         */
        val DIFF_CALLBACK: DiffUtil.ItemCallback<Person> = object : DiffUtil.ItemCallback<Person>() {
            override fun areItemsTheSame(oldItem: Person,
                                         newItem: Person): Boolean {
                return oldItem.personUid == newItem.personUid
            }

            override fun areContentsTheSame(oldItem: Person,
                                            newItem: Person): Boolean {
                return oldItem == newItem
            }
        }
    }
}
